src/core/on_internal_error.cc

```diff
@@ -46,8 +45,11 @@ bool seastar::set_abort_on_internal_error(bool do_abort) noexcept {
     return abort_on_internal_error.exchange(do_abort);
 }
 
-void internal::increase_internal_errors_counter() noexcept{
-    seastar::engine()._internal_errors++;
+namespace seastar::internal {
+thread_local uint64_t internal_errors = 0;
+void increase_internal_errors_counter() noexcept {
+    internal_errors++;
```

Comment:
This can produce really bad code if the compiler thinks the variable can be declared elsewhere.

I don't think this is the case here, but please check the increment translates to one instruction and not a guard variable check followed by initialization.

The antidote is to declare the variable as constinit in .hh; see eaa00e761fc5f6694fde58e72e68ffd62c6a3982 for an example.

(not that it matters since internal errors should be rare, but...)

Reply:
clang 20.1.7 (scylla dbuild)

```
0000000000000010 <_ZN7seastar8internal32increase_internal_errors_counterEv>:
  10:   48 8b 05 00 00 00 00    mov    0x0(%rip),%rax        # 17 <_ZN7seastar8internal32increase_internal_errors_counterEv+0x7>
                        13: R_X86_64_GOTTPOFF   _ZN7seastar12local_engineE-0x4
  17:   64 48 8b 00             mov    %fs:(%rax),%rax
  1b:   48 ff 80 00 17 00 00    incq   0x1700(%rax)
  22:   c3                      ret
  23:   66 2e 0f 1f 84 00 00    cs nopw 0x0(%rax,%rax,1)
  2a:   00 00 00 
  2d:   0f 1f 00                nopl   (%rax)

```

Reply:
Can `increase_internal_errors_counter` be inlined resulting in a single instruction?

Reply:
I was curious enough to check

Reply:
AND :question: